c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -1210,6 +1212,1279 @@ namespace NativeStreamSource {
+    if (!enqueueResult) {
+      // 5.1.  Perform
+      // TransformStreamErrorWritableAndUnblockWrite(stream, enqueueResult.[Value]).
+      RootedValue resultValue(cx);
+      if (!JS_GetPendingException(cx, &resultValue)) {
+        return false;
+      }
+      JS_ClearPendingException(cx);
+
+      if (!TransformStream::ErrorWritableAndUnblockWrite(cx, stream, resultValue)) {
+        return false;
+      }
+
+      //     2.  Throw stream.[readable].[storedError].
+      RootedValue storedError(cx, JS::ReadableStreamGetStoredError(cx, readable));
+      JS_SetPendingException(cx, storedError);
+      return false;
+    }
+
+    // 6.  Let backpressure be ReadableStreamDefaultControllerHasBackpressure(readableController).
+    // (Inlined)
+    bool backpressure = !JS::ReadableStreamControllerShouldCallPull(cx, readableController);
+
+    // 7.  If backpressure is not stream.[backpressure],
+    if (backpressure != TransformStream::backpressure(stream)) {
+      //     1.  Assert: backpressure is true.
+      MOZ_ASSERT(backpressure);
+
+      //     2.  Perform ! TransformStreamSetBackpressure(stream, true).
+      TransformStream::SetBackpressure(cx, stream, true);
+    }
+
+    return true;
+  }
+
+  /**
+   * TransformStreamDefaultControllerTerminate
+   */
+  bool Terminate(JSContext* cx, HandleObject controller) {
+    MOZ_ASSERT(is_instance(controller));
+
+    // 1.  Let stream be controller.[stream].
+    RootedObject stream(cx, ::TransformStreamDefaultController::stream(controller));
+
+    // 2.  Let readableController be stream.[readable].[controller].
+    RootedObject readable(cx, TransformStream::readable(stream));
+    RootedObject readableController(cx, JS::ReadableStreamGetController(cx, readable));
+    MOZ_ASSERT(readableController);
+
+    // 3.  Perform ! [ReadableStreamDefaultControllerClose](readableController).
+    // Note: in https://github.com/whatwg/streams/pull/1029, the spec was changed to
+    // make ReadableStreamDefaultControllerClose (and -Enqueue) return early if
+    // ReadableStreamDefaultControllerCanCloseOrEnqueue is false. SpiderMonkey hasn't been
+    // updated accordingly, so it'll throw an exception instead.
+    // To avoid that, we do the check explicitly. While that also throws an exception, we can
+    // just clear it and move on.
+    if (JS::CheckReadableStreamControllerCanCloseOrEnqueue(cx, readableController, "close")) {
+      if (!JS::ReadableStreamClose(cx, readable)) {
+        return false;
+      }
+    } else {
+      JS_ClearPendingException(cx);
+    }
+
+    // 4.  Let error be a `[TypeError](https://tc39.es/ecma262/#sec-native-error-types-used-in-this-standard-typeerror)` exception indicating that the stream has been terminated.
+    RootedValue error(cx);
+    JS_ReportErrorLatin1(cx, "The TransformStream has been terminated");
+    if (!JS_GetPendingException(cx, &error)) {
+      return false;
+    }
+    JS_ClearPendingException(cx);
```

Comment:
This sequence is somewhat confusing at least to me: we report an error (which creates an exception on the context) then clear it? Perhaps a comment here to describe intent would help...